Mathlib/GroupTheory/GroupAction/Hom.lean

```diff
@@ -1,212 +1,524 @@
-  map_smul := MulActionHom.map_smul'
+  map_smulₛₗ := MulActionHom.map_smul'
 
 initialize_simps_projections MulActionHom (toFun → apply)
 
 namespace MulActionHom
 
-variable {M M' X Y}
+variable {φ X Y}
+variable {F : Type*}
 
 /- porting note: inserted following def & instance for consistent coercion behaviour,
 see also Algebra.Hom.Group -/
-/-- Turn an element of a type `F` satisfying `SMulHomClass F M X Y` into an actual
-`MulActionHom`. This is declared as the default coercion from `F` to `MulActionHom M X Y`. -/
+/-- Turn an element of a type `F` satisfying `MulActionSemiHomClass F φ X Y`
+  into an actual `MulActionHom`.
+  This is declared as the default coercion from `F` to `MulActionSemiHom φ X Y`. -/
 @[coe]
-def _root_.SMulHomClass.toMulActionHom [SMul M X] [SMul M Y] [SMulHomClass F M X Y] (f : F) :
-    X →[M] Y where
-   toFun := FunLike.coe f
-   map_smul' := map_smul f
+def _root_.MulActionSemiHomClass.toMulActionHom [MulActionSemiHomClass F φ X Y] (f : F) :
+    X →ₑ[φ] Y where
+  toFun := FunLike.coe f
+  map_smul' := map_smulₛₗ f
 
-/-- Any type satisfying `SMulHomClass` can be cast into `MulActionHom` via
-  `SMulHomClass.toMulActionHom`. -/
-instance [SMul M X] [SMul M Y] [SMulHomClass F M X Y] : CoeTC F (X →[M] Y) :=
-  ⟨SMulHomClass.toMulActionHom⟩
+/-- Any type satisfying `MulActionSemiHomClass` can be cast into `MulActionHom` via
+  `MulActionHomSemiClass.toMulActionHom`. -/
+instance [MulActionSemiHomClass F φ X Y] : CoeTC F (X →ₑ[φ] Y) :=
+  ⟨MulActionSemiHomClass.toMulActionHom⟩
 
 variable (M' X Y F) in
 /-- If Y/X/M forms a scalar tower, any map X → Y preserving X-action also preserves M-action. -/
 def _root_.IsScalarTower.smulHomClass [MulOneClass X] [SMul X Y] [IsScalarTower M' X Y]
-    [SMulHomClass F X X Y] : SMulHomClass F M' X Y where
-  map_smul f m x := by
-    rw [← mul_one (m • x), ← smul_eq_mul, map_smul, smul_assoc, ← map_smul, smul_eq_mul, mul_one]
+    [MulActionHomClass F X X Y] : MulActionHomClass F M' X Y where
+  map_smulₛₗ f m x := by
+    rw [← mul_one (m • x), ← smul_eq_mul, map_smul, smul_assoc, ← map_smul,
+      smul_eq_mul, mul_one, id_eq]
 
 protected theorem map_smul (f : X →[M'] Y) (m : M') (x : X) : f (m • x) = m • f x :=
   map_smul f m x
 #align mul_action_hom.map_smul MulActionHom.map_smul
 
 @[ext]
-theorem ext {f g : X →[M'] Y} : (∀ x, f x = g x) → f = g :=
+theorem ext [MulActionSemiHomClass F φ X Y] {f g : F} :
+    (∀ x, f x = g x) → f = g :=
   FunLike.ext f g
 #align mul_action_hom.ext MulActionHom.ext
 
-theorem ext_iff {f g : X →[M'] Y} : f = g ↔ ∀ x, f x = g x :=
+theorem ext_iff [MulActionSemiHomClass F φ X Y] {f g : F} : f = g ↔ ∀ x, f x = g x :=
   FunLike.ext_iff
 #align mul_action_hom.ext_iff MulActionHom.ext_iff
 
-protected theorem congr_fun {f g : X →[M'] Y} (h : f = g) (x : X) : f x = g x :=
+protected theorem congr_fun [MulActionSemiHomClass F φ X Y] {f g : F} (h : f = g) (x : X) :
+    f x = g x :=
   FunLike.congr_fun h _
```

Comment:
I'm not sure I like these changes: really `ext_iff` and `congr_fun` should always be `FunLike.ext_iff` and `FunLike.congr_fun`, this is just a reimplementation for backwards compatibility. (And on #8386 these definitions will cause a "unused variable" error, since `FunLike` will no longer be part of `HomClass`es.)

Reply:
Sorry, I don't understand the comment.
Should I just have `congr_fun` and `ext_iff` for the MulActionHom types, and not the use the MulActionSemiHomClass?